designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+
+To workaround the limitation of round-trippability, we are proposing a rename of both the API group and Kind (`NodePool`) that the CRDs exist within. This allows both CRDs to exist alongside each other simultaneously and gives users a natural migration path to move through.
+
+### Migration Path
+
+#### Aggressive Scale-Down
+
+1. Spin up a new `v1beta1/NodePool` and `v1beta1/NodeClass` resource matching their `v1alpha5` counterparts
+2. Delete the `v1alpha5/Provisioner` and `v1alpha5/AWSNodeTemplate` with [cascading foreground deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#foreground-deletion) to delete the nodes and have Karpenter roll the nodes to `v1beta1`
+
+#### Slow Scale-Down
+
+1. Spin up a new `v1beta1/NodePool` and `v1beta1/NodeClass` resource matching their `v1alpha5` counterparts
+2. Set the `spec.limits` on your `v1alpha5/Provisioner` resource to `cpu=0` to stop provisioning
+3. Manually delete Nodes one-by-one to have Karpenter roll the nodes over to `v1beta1`
+
+## APIs
+
+To help clearly define where configuration should live within Karpenter’s API, we should define the logical boundary between each Kind in the project.
+
+1. `NodePool`
+    1. Neutral Node configuration-based fields that affect the **compatibility between Nodes and Pods during scheduling** (e.g. requirements, taints, labels)
+    2. Neutral behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making
+2. `NodeClass`
+    1. Cloudprovider-specific Node configuration-based fields that affect launch and bootstrap process for that Node including: configuring startup scripts (including kubelet configuration), volume mappings, metadata settings, etc.
+    2. Cloudprovider-specific behavior-based fields for configuring Karpenter’s scheduling and deprovisioning decision-making (e.g. interruption-based disruption, allocation strategy)
+3. `NodeClaim`
+    1. A Karpenter management object that fully manages the lifecycle of a single node including: configuring and launching the node, monitoring the node health (including disruption conditions), and handling the deprovisioning and termination of the node
+
+### `karpenter.sh/NodePool`
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClass:
+        name: default
+        kind: NodeClass
+        apiVersion: compute.k8s.aws/v1beta1
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+      resources:
+        requests:
+          cpu: "1"
+          memory: "100Mi"
```

Comment:
What would resource requests mean here?

Reply:
These requests are used by the NodeClaim to understand the minimum set of pod requests that are required for the node to be launched. By default, we aren't going to allow this value to be set in the `NodePool` (but the field will still be available).

I could imagine that this field could be available in the future for the top-level `NodePool` to be a secondary mechanism to define the minimum set of requests required by a NodeClaim that Karpenter launches from that pool similar to the current `Gt`/`Lt` operators.